### DISCUSSION TITLE
How to GridFunctionCoefficient::Eval on a Different Mesh?

### DISCUSSION THREAD

**Discussion by j-signorelli** [mfem/mfem#3889] [Q&A]
Hi,

I am in the process of implementing a regression test for restarting simulations for code I wrote that relies heavily on MFEM. I have a ParGridFunction `u_0` that I write out using VisitDataCollection, and then I import that outputted ParGridFunction into a new ParGridFunction `u_r`. To ensure that the functionality is working, I am trying essentially to do:

    GridFunctionCoefficient u_0_coeff(u_0);
    u_r->ComputeL2Error(u_0_coeff);

However, although the meshes associated with both ParGridFunction's are physically the same, they are different objects (pointing to different locations in memory). This leads to the following bug:

    Verification failed: (GetLastOperation() == Mesh::REFINE) is false:
    --> 
    ... in function: const mfem::CoarseFineTransformations& mfem::Mesh::GetRefinementTransforms()

 as I traced this through: `ParGridFunction::ComputeL2Error` --> `GridFunction::ComputeL2Error` --> `GridFunction::ComputeLpError` --> `GridFunctionCoefficient::Eval` --> `Mesh::GetRefinementTransforms`.

In `GridFunctionCoefficient::Eval`, it seems to me that there is first a check to see if the meshes are pointing to the same location in memory, and if not, then it seems to presume that one mesh is a refined version of another?

Nonetheless: I believe that this error is occurring because `u_r` has no refinement done on it. Is this true? And if so, is there anyway that I can evaluate a GridFunctionCoefficient on a "different" (in memory) mesh?

Thank you for any assistance.

**Comment by vladotomov** [comment]
Hi @j-signorelli,

Does it work if you comment out this verification? I remember we've hit this in the past, and I think this check assumes that different meshes can only be used for LOR operations. @pazner can probably comment better.

**Reply by j-signorelli** [reply]
Hi @vladotomov 

Thank you for your response. I ended up just using `Vector::DistanceTo` for each rank's ParGridFunction and just ensured that it is ~0 as a workaround.

However, I did just check this by commenting out the verification and unfortunately still got this error on every rank:

    MFEM abort: Don't know how to construct CoarseFineTransformations for geom = 3
        ... in function: const mfem::CoarseFineTransformations& mfem::Mesh::GetRefinementTransforms()
